central/platform/reprocessor/reprocessor_impl.go

```diff
@@ -0,0 +1,187 @@
+package reprocessor
+
+import (
+	"context"
+	"sync/atomic"
+
+	alertDS "github.com/stackrox/rox/central/alert/datastore"
+	alertutils "github.com/stackrox/rox/central/alert/utils"
+	deploymentDS "github.com/stackrox/rox/central/deployment/datastore"
+	platformmatcher "github.com/stackrox/rox/central/platform/matcher"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/search"
+)
+
+const batchSize = 5000
+
+var (
+	log = logging.LoggerForModule()
+
+	reprocessorCtx = sac.WithAllAccess(context.Background())
+
+	query = search.NewQueryBuilder().AddNullField(search.PlatformComponent).ProtoQuery()
+)
+
+type platformReprocessorImpl struct {
+	alertDatastore      alertDS.DataStore
+	deploymentDatastore deploymentDS.DataStore
+	platformMatcher     platformmatcher.PlatformMatcher
+
+	stopSignal concurrency.Signal
+	// isStarted will make sure only one reprocessing routine runs for an instance of reprocessor
+	isStarted atomic.Bool
+}
+
+func New(alertDatastore alertDS.DataStore,
+	deploymentDatastore deploymentDS.DataStore,
+	platformMatcher platformmatcher.PlatformMatcher) PlatformReprocessor {
+
+	return &platformReprocessorImpl{
+		alertDatastore:      alertDatastore,
+		deploymentDatastore: deploymentDatastore,
+		platformMatcher:     platformMatcher,
+		stopSignal:          concurrency.NewSignal(),
+	}
+}
+
+func (pr *platformReprocessorImpl) Start() {
+	swapped := pr.isStarted.CompareAndSwap(false, true)
+	if !swapped {
+		log.Error("Platform reprocessor was already started")
+		return
+	}
+	go pr.runReprocessing()
+}
+
+func (pr *platformReprocessorImpl) Stop() {
+	if !pr.isStarted.Load() {
+		log.Error("Platform reprocessor not started")
+	}
+	pr.stopSignal.Signal()
+}
+
+func (pr *platformReprocessorImpl) runReprocessing() {
+	needsReprocessing, err := pr.needsReprocessing()
+	if err != nil {
+		log.Errorf("Error determining if platform components need reporcessing: %v", err)
+		return
+	}
+
+	if !needsReprocessing {
+		log.Info("Platform components up to date, skipping reprocessing")
+		return
+	}
+
+	err = pr.reprocessAlerts()
+	if err != nil {
+		log.Errorf("Error reprocessing alerts with platform rules: %v", err)
+		return
+	}
+
+	err = pr.reprocessDeployments()
+	if err != nil {
+		log.Errorf("Error reprocessing deployments with platform rules: %v", err)
+		return
+	}
+}
+
+func (pr *platformReprocessorImpl) needsReprocessing() (bool, error) {
+	q := query.CloneVT()
+	q.Pagination = &v1.QueryPagination{
+		Limit: 1,
+	}
+
+	alerts, err := pr.alertDatastore.GetByQuery(reprocessorCtx, q)
+	if err != nil {
+		return false, err
+	}
+	deployments, err := pr.deploymentDatastore.SearchRawDeployments(reprocessorCtx, q)
+	if err != nil {
+		return false, err
+	}
+	return len(alerts) > 0 || len(deployments) > 0, nil
```

Comment:
What are the odds that one of these need reprocessed and the other doesn't?  As it is written we could return after the alerts check and not do the deployment check if the alerts need reprocessed.

OR should we have separate checks and only reprocess the ones that need it?

Reply:
I don't think it is likely that only one of them need reprocessed, but it can be possible if all deployments were updated just before an ACS upgrade. Then the other reprocessor will re-insert them and they won't need reprocessing. I think separate checks for both would be better.